src/sys/windows/selector.rs

```diff
@@ -166,16 +186,61 @@ impl Registration {
     /// Eventually this function will probably also be modified to handle the
     /// `level()` polling option.
     pub fn push_event(&mut self, set: EventSet, events: &mut Vec<IoEvent>) {
+        trace!("push_event; token={:?}; set={:?}; opts={:?}", self.token, set, self.opts);
+
         // If we're not actually interested in any of these events,
         // discard the event, and then if we're actually delivering an event we
         // stop listening if it's also a oneshot.
         let set = self.interest & set;
+
         if set != EventSet::none() {
-            events.push(IoEvent::new(set, self.token));
+            let event = IoEvent::new(set, self.token);
+
+            if self.opts.is_edge() {
+                events.push(event);
+
+                if self.opts.is_oneshot() {
+                    trace!("deregistering because of oneshot");
+                    self.interest = EventSet::none();
+                }
+            } else {
```

Comment:
I'm not quite sure if there are well defined semantics for both edge and level, but perhaps this should be `if opts.is_edge()` followed by `if opts.is_level()` instead of using `else` to make them disjoint?